File: web-codingbat-solutions/coding-bat-java/Calentamiento2.java
/* 
 * Problemas de calentamiento de bucles de strings y arrays
 */

public class Calentamiento2 {
    /* 
    Dada una cadena y un entero n no negativo, retorna una cadena más con el número de copia indicado por n.

    copiaCadena("Hi", 2) → "HiHi"
    copiaCadena("Hi", 3) → "HiHiHi"
    copiaCadena("Hi", 1) → "Hi"
    */

    public String copiaCadena(String cadena, int n) {
        String nuevaCadena = "";
        
        while (n > 0) {
            nuevaCadena += cadena;
            n--;
        }
        return nuevaCadena;
    }

    /* 
    Dada una cadena, retorna verdadero si la primera aparición de "x" en la cadena es seguida inmediatamente por otra "x". 

    dobleX("axxbb") → true
    dobleX("axaxax") → false
    dobleX("xxxxx") → true
    */
    boolean dobleX(String cadena) {
        for (int i = 0; i < cadena.length(); i++) {
            if (i < cadena.length()-1) {
                char ac = cadena.charAt(i);
                char sc = cadena.charAt(i+1);
                if (ac == 'x') {
                    if (sc == 'x') {
                        return true;
                    } else {
                        break;
                    }
                }
            }
        }
        return false;
    }
    /* 
    Dada una cadena, retorna la cantidad de las veces que los dos últimos caracteres de la cadena aparace en la cadena, por lo que "hixxxhi" produce 1 (no contaremos la subcadena final). 

    ultimosDos("hixxhi") → 1
    ultimosDos("xaxxaxaxx") → 1
    ultimosDos("axxxaaxx") → 2
    */
    public int ultimosDos(String cadena) {
        if (cadena.length() < 2) return 0;

        String ultDos = cadena.substring(cadena.length()-2);

        int i = 0;
        int contador = 0;

        while (i < cadena.length()-2) {
            if (cadena.substring(i, i+2).equals(ultDos)) {
                contador++;
            }
            i++;
        }
        return contador;
    }

    /* 
    Dada una matriz de enteros, retorna verdadero si la secuencia de números 1,2,3 aparece en la matriz.

    matriz123([1, 1, 2, 3, 1]) → true
    matriz123([1, 1, 2, 4, 1]) → false
    matriz123([1, 1, 2, 1, 2, 3]) → true
    */
    public boolean matriz123(int[] nums) {
        boolean secuencia = false;
        
        for (int i = 0; i < nums.length; i++) {
            if (i < nums.length-2) {
                if (nums[i] == 1 && nums[i+1] == 2 && nums[i+2] == 3) {
                    secuencia = true; 
                }
            }
        }
        if (secuencia) return true;
        return false;
    }

    /* 
    Dada una cadena, retorna una cadena con los caracteres en los indices 0, 1, 4, 5, 8, 9 por lo tanto "Kittens" produce "Kien".

    paresAlternos("kitten") → "kien"
    paresAlternos("Chocolate") → "Chole"
    paresAlternos("CodingHorror") → "Congrr"
    */
    public String paresAlternos(String cadena) {
        String nuevaCadena = "";
        
        for (int i = 0; i < cadena.length(); i+=4) {
            if (cadena.length() > 1) {
                nuevaCadena += cadena.charAt(i);
                if (i < cadena.length()-1) {
                    nuevaCadena += cadena.charAt(i+1);
                }
            } else {
                return cadena;
            }
        }
        return nuevaCadena;
    }

    /* 
    Dada una matriz de enteros, diremos que es triple si un mismo valor aparece 3 veces seguidas en la matriz. Retorna verdadero si la matriz no contiene ningún triple.

    noTres([1, 1, 2, 2, 1]) → true
    noTres([1, 1, 2, 2, 2, 1]) → false
    noTres([1, 1, 1, 2, 2, 2, 1]) → false
    */
    public boolean noTres(int[] nums) {
        boolean tres = false;
        
        for (int i = 0; i < nums.length; i++) {
            if (i < nums.length-2) {
                if (nums[i] == nums[i+1] && nums[i+1] == nums[i+2]) {
                    tres = true;
                }
            }
        }
        
        if (tres) return false;
        return true;
    }

    /* 
    Dada una cadena y un entero n no negativo, diremos que el frente de la cadena son los 3 primeros caracteres, o lo que contenga si la longitud es menor a 3. Retorna n copies del frente.

    copiasFrente("Chocolate", 2) → "ChoCho"
    copiasFrente("Chocolate", 3) → "ChoChoCho"
    copiasFrente("Abc", 3) → "AbcAbcAbc"
    */
    public String copiasFrente(String cadena, int n) {
        String nuevaCadena = "";
        int contador = 0;
        
        if (n == 0) {
            return "";
        }
        
        while (contador < n) {
            if (cadena.length() > 2) {
                nuevaCadena += cadena.substring(0, 3);
                contador++;  
            } else {
                nuevaCadena += cadena;
                contador++;
            }
        }
        return nuevaCadena;
    }

    /* 
    Dada una cadena, retorna una nueva cadena formada por todos los caracteres pares, por lo que "Hello" produce "Hlo".

    cadenaBits("Hello") → "Hlo"
    cadenaBits("Hi") → "H"
    cadenaBits("Heeololeo") → "Hello"
    */
    public String cadenaBits(String cadena) {
        String nuevaCadena = "";
        int i = 0;
        
        while (i < cadena.length()) {
            nuevaCadena += cadena.charAt(i);
            i+=2;
        }
        return nuevaCadena;
    }

    /* 
    Dada una matriz de enteros, retorna el número de veces que el 9 se encuentra en la matriz.

    cuentaMatriz9([1, 2, 9]) → 1
    cuentaMatriz9([1, 9, 9]) → 2
    cuentaMatriz9([1, 9, 9, 3, 9]) → 3
    */
    public int cuentaMatriz9(int[] nums) {
        int contador = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 9) {
                contador++;
            }
        }
        return contador;
    }

    /* 
    Dadas dos cadena a y b, retorna el numero de veces en las que una de las cadenas contiene la misma subcadena de longitud 2. Por lo tanto, "xxcaazz" dan como resultado 3, ya que las subcadenas "xx", "aa" y "az" aparecen en el mismo lugar en ambas cadenas. 

    cadenaPartida("xxcaazz", "xxbaaz") → 3
    cadenaPartida("abc", "abc") → 2
    cadenaPartida("abc", "axc") → 0
    */
    public int cadenaPartida(String a, String b) {
        int contador = 0;
        int longitud = Math.min(a.length(), b.length());
        
        for (int i = 0; i < longitud; i++) {
            if (i < longitud-1) {
                String sub = a.substring(i, i+2);
                if (sub.equals(b.substring(i, i+2))) {
                    contador++;
                }
            }
        }
        return contador;
    }

    /* 
    Supongamos que la cadena "yak" es desafortunada. Dada una cadena, retorna una versión en la que se eliminen todos los "yak", per la "a" puede ser cualquier carácter. Las cadenas "yak" no se superpondrán.

    cadenaYak("yakpak") → "pak"
    cadenaYak("pakyak") → "pak"
    cadenaYak("yak123ya") → "123ya"
    */
    public String cadenaYak(String cadena) {
        String nuevaCadena = "";
        
        for (int i = 0; i < cadena.length(); i++) {
            if (i < cadena.length()-2 && cadena.charAt(i) == 'y' && cadena.charAt(i+2) == 'k') {
                i = i + 2;
            } else {
                nuevaCadena += cadena.charAt(i);
            }
        }
        return nuevaCadena;
    }

    /* 
    Dada una matriz de enteros, retorna verdadera si esta contiene un patron 2, 7, 1: un valor seguido del otro más 5, seguido del valor menos 1. Además, el 271 cuenta incluso si el "1" difiere en 2 o menos del valor correcto.

    tiene271([1, 2, 7, 1]) → true
    tiene271([1, 2, 8, 1]) → false
    tiene271([2, 7, 1]) → true
    */
    public boolean tiene271(int[] nums) {
        for (int i = 0; i < nums.length-2; i++) {
            int siete = nums[i]+5;
            int dos = Math.abs(nums[i+2] - (siete - 6));
            
            if (nums[i+1] == siete && dos <= 2) {
                return true;
            }
        }
        return false;
    }

    /* 
    Cuenta el numero de "xx" en la cadena dada. Supongamos que se permite la superposición, por lo que "xxx" contiene 2 "xx".

    cuentaXX("abcxx") → 1
    cuentaXX("xxx") → 2
    cuentaXX("xxxx") → 3
    */
    public int cuentaXX(String cadena) {
        int contador = -1;
        
        for (int i = 0; i < cadena.length(); i++) {
            if (Character.isWhitespace(cadena.charAt(i))) {
                contador++;
                break;
            }
            if (cadena.charAt(i) == 'x') {
                contador++;
            }
        }
        if (contador == -1) return 0;
        return contador;
    }

    /* 
    Dada una cadena no vacía como "Code", devuelve una cadena como "CCoCodCode".

    cadenaExplosion("Code") → "CCoCodCode"
    cadenaExplosion("abc") → "aababc"
    cadenaExplosion("ab") → "aab"
    */
    public String cadenaExplosion(String cadena) {
        String nuevaCadena = "";
        int contador = 1;

        for (int i = 0; i < cadena.length(); i++) {
            for (int j = 0; j < contador; j++) {
                nuevaCadena += cadena.charAt(j);
            }
            contador++;
        }
        return nuevaCadena;
    }

    /* 
    Dada una matriz de enteros, retorna verdadero si uno de los primeros 4 elementos en la matriz es un 9. La longitud de la matriz puede ser menor que 4.

    frenteMatriz9([1, 2, 9, 3, 4]) → true
    frenteMatriz9([1, 2, 3, 4, 9]) → false
    frenteMatriz9([1, 2, 3, 4, 5]) → false
    */
    public boolean frenteMatriz9(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (i < 4) {
                if (nums[i] == 9) {
                    return true;
                }
            }
        }
        return false;
    }

    /* 
    Dada una cadena, retorna una version en la que se han eliminado todas las "x". Excepto que no se debe eliminar una "x" al principio o al final.
    
    cadenaX("xxHxix") → "xHix"
    cadenaX("abxxxcd") → "abcd" 
    cadenaX("xabxxxcdx") → "xabcdx"
    */
    public String cadenaX(String cadena) {
        String nuevaCadena = "";

        if (cadena.length() > 1) {
            nuevaCadena += cadena.charAt(0);
            for (int i = 0; i < cadena.length(); i++) {
                if (i > 0 && i < cadena.length()-1) {
                    if (cadena.charAt(i) != 'x') {
                        nuevaCadena += cadena.charAt(i);
                    }
                }
            }
            return nuevaCadena + cadena.charAt(cadena.length()-1);
        }
        return cadena;
    }

    /* 
    Dada una matriz de enteros, retorna el numero de veces que dos 6 están uno al lado del otro en la matriz. También cuenta las instancias en las que el segundo "6" es en realidad un 7. 

    matriz667([6, 6, 2]) → 1
    matriz667([6, 6, 2, 6]) → 1
    matriz667([6, 7, 2, 6]) → 1
    */
    public int matriz667(int[] nums) {
        int contador = 0;

        for (int i = 0; i < nums.length; i++) {
            if (i < nums.length-1) {
                if (nums[i] == 6 && (nums[i+1] == 6 || nums[i+1] == 7)) {
                    contador++;
                }
            }
        }
        return contador;
    }
}

File: web-codingbat-solutions/coding-bat-python/calentamiento2.py
"""
Problemas de listas/cuerdas de calentamiento medio con bucles
"""

"""
Dada una cadena y un entero no negativo n, devuelve una cadena más grande que contiene n copias de la cadena original.

repetir_cadena('Hi', 2) → 'HiHi'
repetir_cadena('Hi', 3) → 'HiHiHi'
repetir_cadena('Hi', 1) → 'Hi'
"""
def repetir_cadena(str, n):
    return str * n

"""
Dada una cadena no vacía como "Código", se devuelve una cadena como "CCoCodCode".

cadena_explosion('Código') → 'CCoCodCode'
cadena_explosion('abc') → 'aababc'
cadena_explosion('ab') → 'aab'
"""
def string_splosion(cadena):
    nueva_cadena = ""

    contador = 1
    for i in range(len(cadena)):
        for j in range(0, contador, 1):
            nueva_cadena += cadena[j]
        contador += 1
    return nueva_cadena

"""
Dado un array de enteros, devuelve Verdadero si uno de los primeros 4 elementos del array es un 9. La longitud del array puede ser menor que 4.

matriz_frente_9([1, 2, 9, 3, 4]) → Verdadero
matriz_frente_9([1, 2, 3, 4, 9]) → Falso
matriz_frente_9([1, 2, 3, 4, 5]) → Falso
"""
def matriz_frente_9(nums):
    for i in range(len(nums)):
        if i < 4:
            if nums[i] == 9:
                return True
    return False

"""
Dada una cadena y un entero no negativo n, diremos que el inicio de la cadena son los 3 primeros caracteres, o el que esté presente si la cadena tiene una longitud menor a 3. Devuelve n copias del inicio;

repetir_frente('Chocolate', 2) → 'ChoCho'
repetir_frente('Chocolate', 3) → 'ChoChoCho'
repetir_frente('Abc', 3) → 'AbcAbcAbc'
"""
def repetir_frente(str, n):
    return str[0:3] * n

"""
Dada una cadena, devuelve el número de veces que una subcadena de longitud 2 aparece en la cadena, y también como los dos últimos caracteres de la cadena; por lo tanto, "hixxxhi" da como resultado 1 (no se contará la subcadena final).

ultimos2('hixxhi') → 1
ultimos2('xaxxaxaxx') → 1
ultimos2('axxxaaxx') → 2
"""
def ultimos2(cadena):
    if len(cadena) < 2: 
        return 0

    i = 0
    contador = 0

    while (i < len(cadena)-2):
        if cadena[i:i+2] == cadena[len(cadena)-2:]:
            contador += 1
        i += 1

    return contador

"""
Dado un array de enteros, devuelve Verdadero si la secuencia de números 1, 2, 3 aparece en algún lugar del array.

matriz123([1, 1, 2, 3, 1]) → Verdadero
matriz123([1, 1, 2, 4, 1]) → Falso
matriz123([1, 1, 2, 1, 2, 3]) → Verdadero
"""
def matriz123(nums):
    secuencia = False

    for i in range(len(nums)): 
        if i < len(nums)-2:
            if nums[i] == 1 and nums[i+1] == 2 and nums[i+2] == 3:
                secuencia = True 
    if secuencia: 
        return True
    return False

"""
Dada una cadena, devuelve una nueva cadena compuesta por todos los demás caracteres, comenzando por el primero. Por lo tanto, "Hola" da como resultado "Hlo".

pedazo_cadena('Hola') → 'Hlo'
pedazo_cadena('Hi') → 'H'
pedazo_cadena('Heeololeo') → 'Hello'
"""
def pedazo_cadena(cadena):
    nueva_cadena = ""
    i = 0

    while (i < len(cadena)):
        nueva_cadena += str[i]
        i += 2
    return nueva_cadena

"""
Dado un array de enteros, devuelve el número de 9 en el array.

matriz_cuenta9([1, 2, 9]) → 1
matriz_cuenta9([1, 9, 9]) → 2
matriz_cuenta9([1, 9, 9, 3, 9]) → 3
"""
def matriz_cuenta9(nums):
    count = 0

    for i in range(len(nums)):
        if nums[i] == 9:
            count += 1
    return count

"""
Dadas dos cadenas, a y b, devuelve el número de posiciones donde contienen la misma subcadena de longitud 2. Por lo tanto, "xxcaazz" y "xxbaaz" dan como resultado 3, ya que las subcadenas "xx", "aa" y "az" aparecen en el mismo lugar en ambas cadenas.

cadena_fosforo('xxcaazz', 'xxbaaz') → 3
cadena_fosforo('abc', 'abc') → 2
cadena_fosforo('abc', 'axc') → 0
"""
def cadena_fosforo(a, b):
    contador = 0;
    longitud = min(len(a), len(b))

    for i in range(longitud):
        if i < longitud-1:
            sub = a[i:i+2]
            if (sub == b[i:i+2]):
                contador += 1
    return contador